File: TypeScript/Math/get-missing-int-in-slice.ts
function main(): void {
    // Create a slice of ints
    let randomValues: number[] = [1, 2, 3, 5, 6, 7, 8, 9, 10]
    // Get the missing int in the slice
    console.log(getMissingIntInSlice(randomValues))
}

main()

// Get the missing int in a slice of ints
function getMissingIntInSlice(slice: number[]): number {
    let smallestInt: number = getSmallestIntInSlice(slice)
    let biggestInt: number = getBiggestIntInSlice(slice)
    let missingInt: number = slice[0]
    for (let i: number = smallestInt; i <= biggestInt; i++) {
        if (!doesSliceContainElement(slice, i)) {
            missingInt = i
        }
    }
    return missingInt
}

// Check if a given slice contains a given element
function doesSliceContainElement(slice: number[], element: number): boolean {
    for (let i: number = 0; i < slice.length; i++) {
        if (slice[i] === element) {
            return true
        }
    }
    return false
}

// Get the smallest int in a slice of ints
function getSmallestIntInSlice(slice: number[]): number {
    let smallestInt: number = slice[0]
    for (let i: number = 1; i < slice.length; i++) {
        if (slice[i] < smallestInt) {
            smallestInt = slice[i]
        }
    }
    return smallestInt
}

// Get the biggest int in a slice of ints
function getBiggestIntInSlice(slice: number[]): number {
    let biggestInt: number = slice[0]
    for (let i: number = 1; i < slice.length; i++) {
        if (slice[i] > biggestInt) {
            biggestInt = slice[i]
        }
    }
    return biggestInt
}

File: Python/Math/get_missing_int_in_slice.py
# Get the missing int in a slice of ints.
def get_missing_int_in_slice(given_slice):
    smallestInt = get_smallest_int_in_slice(given_slice)
    biggestInt = get_biggest_int_in_slice(given_slice)
    missingInt = []
    for tempValue in range(smallestInt, biggestInt):
        if slice_contains(given_slice, tempValue) == False:
            missingInt.append(tempValue)
    return missingInt


# Get the biggest int in a given slice.
def get_biggest_int_in_slice(given_slice):
    return max(given_slice)


# Get the smallest int in a given slice.
def get_smallest_int_in_slice(given_slice):
    return min(given_slice)


# Determine whether or not the array includes a specific element.
def slice_contains(provided_slice, provided_element):
    return provided_element in provided_slice


def main():
    # Create a slice of ints
    randomSlice = [1, 10]
    # Get the missing int in the slice.
    print(get_missing_int_in_slice(randomSlice))


main()
